File: week1_baseline/ruby/10_standard_tool_library/lib/boukensha/tools/file_system.rb
require "fileutils"

module Boukensha
  module Tools
    # FileSystem registers the standard set of file-oriented tools against a
    # registry, all sandboxed to a single root directory.
    #
    # Tools registered:
    #   pwd              — return the working directory
    #   list_directory   — list files and subdirectories at a path
    #   read_file        — read the full contents of a file
    #   write_file       — write (or overwrite) a file
    #   delete_file      — delete a file
    #   search_files     — grep for a pattern across files in the working tree
    #
    # Every path argument the agent supplies is resolved relative to that root.
    # If the resolved path would escape the root (path traversal) the tool
    # returns an error string rather than raising — so the agent sees it and
    # can try something sensible instead.
    #
    # Usage (handled automatically by Boukensha.run / Boukensha.repl when working_dir:
    # is set, but you can call it directly too):
    #
    #   Boukensha::Tools::FileSystem.register(registry, working_dir: "/my/project")
    #
    module FileSystem
      def self.register(registry, working_dir:)
        root = File.expand_path(working_dir)

        # Resolve a relative (or absolute) agent-supplied path inside root.
        # Returns the absolute path on success, or an error String.
        resolve = lambda do |path|
          absolute = File.expand_path(path.to_s, root)
          if absolute == root || absolute.start_with?("#{root}/")
            absolute
          else
            "error: path '#{path}' escapes the working directory"
          end
        end

        oops = ->(msg) { "error: #{msg}" }

        registry.tool "pwd",
          description: "Return the working directory — the root that all file paths are relative to.",
          parameters:  {} do
          root
        end

        registry.tool "list_directory",
          description: "List files and subdirectories at a path relative to the working directory. Defaults to the working directory itself.",
          parameters:  {
            path: { type: "string", description: "Relative path to list (default '.')" }
          } do |path: "."|
          target = resolve.call(path)
          next target if target.start_with?("error:")
          next oops.call("'#{path}' is not a directory") unless File.directory?(target)

          entries = Dir.entries(target)
                       .reject { |e| e == "." || e == ".." }
                       .sort
                       .map { |name| File.directory?(File.join(target, name)) ? "#{name}/" : name }

          entries.empty? ? "(empty)" : entries.join("\n")
        end

        registry.tool "read_file",
          description: "Read and return the full contents of a file. Path is relative to the working directory.",
          parameters:  {
            path: { type: "string", description: "Relative path to the file" }
          } do |path:|
          target = resolve.call(path)
          next target if target.start_with?("error:")
          next oops.call("'#{path}' is not a file") unless File.file?(target)

          File.read(target)
        rescue => e
          oops.call(e.message)
        end

        registry.tool "write_file",
          description: "Write content to a file, creating it (and any missing parent directories) if needed, overwriting if it exists. Path is relative to the working directory.",
          parameters:  {
            path:    { type: "string", description: "Relative path to the file" },
            content: { type: "string", description: "Text content to write" }
          } do |path:, content:|
          target = resolve.call(path)
          next target if target.start_with?("error:")

          FileUtils.mkdir_p(File.dirname(target))
          File.write(target, content)
          rel = target.delete_prefix("#{root}/")
          "ok: wrote #{content.bytesize} bytes to #{rel}"
        rescue => e
          oops.call(e.message)
        end

        registry.tool "delete_file",
          description: "Delete a file. Directories are not deleted. Path is relative to the working directory.",
          parameters:  {
            path: { type: "string", description: "Relative path to the file to delete" }
          } do |path:|
          target = resolve.call(path)
          next target if target.start_with?("error:")
          next oops.call("'#{path}' is not a file") unless File.file?(target)

          File.delete(target)
          "ok: deleted #{path}"
        rescue => e
          oops.call(e.message)
        end

        registry.tool "search_files",
          description: "Search for a text pattern (literal string or Ruby regex) across all files in the working directory tree. Returns matching lines in 'path:line_number:content' format.",
          parameters:  {
            pattern: { type: "string", description: "The text or regex pattern to search for" },
            path:    { type: "string", description: "Subdirectory or file to search within (default '.' = entire working directory)" },
            glob:    { type: "string", description: "File glob to restrict which files are searched, e.g. '*.rb' (default '*')" }
          } do |pattern:, path: ".", glob: "*"|
          target = resolve.call(path)
          next target if target.start_with?("error:")

          search_root = File.file?(target) ? File.dirname(target) : target
          file_glob   = File.file?(target) ? target : File.join(target, "**", glob)

          begin
            regex   = Regexp.new(pattern)
          rescue RegexpError => e
            next oops.call("invalid pattern: #{e.message}")
          end

          matches = []
          Dir.glob(file_glob).sort.each do |file|
            next unless File.file?(file)
            rel = file.delete_prefix("#{root}/")
            File.foreach(file).with_index(1) do |line, lineno|
              matches << "#{rel}:#{lineno}:#{line.chomp}" if line.match?(regex)
            end
          rescue => e
            matches << "#{rel}: error reading file: #{e.message}"
          end

          matches.empty? ? "no matches" : matches.join("\n")
        end

      end # def self.register
    end # FileSystem
  end # Tools
end # Boukensha

File: week1_baseline/python/10_standard_tool_library/boukensha/tools/file_system.py
from __future__ import annotations

import os
import re
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..registry import Registry

# FileSystem registers the standard set of file-oriented tools against a
# registry, all sandboxed to a single root directory.
#
# Tools registered:
#   pwd              -- return the working directory
#   list_directory    -- list files and subdirectories at a path
#   read_file        -- read the full contents of a file
#   write_file       -- write (or overwrite) a file
#   delete_file      -- delete a file
#   search_files     -- grep for a pattern across files in the working tree
#
# Every path argument the agent supplies is resolved relative to that root.
# If the resolved path would escape the root (path traversal) the tool
# returns an error string rather than raising -- so the agent sees it and
# can try something sensible instead.


def register(registry: Registry, *, working_dir: str | Path) -> None:
    root = Path(os.path.abspath(os.path.expanduser(str(working_dir))))

    def resolve(path: str) -> Path | str:
        absolute = Path(os.path.normpath(os.path.join(root, path)))
        if absolute == root or str(absolute).startswith(f"{root}{os.sep}"):
            return absolute
        return f"error: path '{path}' escapes the working directory"

    def oops(msg: str) -> str:
        return f"error: {msg}"

    registry.tool(
        "pwd",
        description="Return the working directory — the root that all file paths are relative to.",
        parameters={},
        block=lambda: str(root),
    )

    def list_directory(*, path: str = ".") -> str:
        target = resolve(path)
        if isinstance(target, str):
            return target
        if not target.is_dir():
            return oops(f"'{path}' is not a directory")

        entries = sorted(f"{p.name}/" if p.is_dir() else p.name for p in target.iterdir())
        return "\n".join(entries) if entries else "(empty)"

    registry.tool(
        "list_directory",
        description=(
            "List files and subdirectories at a path relative to the working directory. "
            "Defaults to the working directory itself."
        ),
        parameters={"path": {"type": "string", "description": "Relative path to list (default '.')"}},
        block=list_directory,
    )

    def read_file(*, path: str) -> str:
        target = resolve(path)
        if isinstance(target, str):
            return target
        if not target.is_file():
            return oops(f"'{path}' is not a file")
        try:
            return target.read_text()
        except (OSError, UnicodeDecodeError) as e:
            return oops(str(e))

    registry.tool(
        "read_file",
        description="Read and return the full contents of a file. Path is relative to the working directory.",
        parameters={"path": {"type": "string", "description": "Relative path to the file"}},
        block=read_file,
    )

    def write_file(*, path: str, content: str) -> str:
        target = resolve(path)
        if isinstance(target, str):
            return target
        try:
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_text(content)
            rel = target.relative_to(root)
            return f"ok: wrote {len(content.encode())} bytes to {rel}"
        except OSError as e:
            return oops(str(e))

    registry.tool(
        "write_file",
        description=(
            "Write content to a file, creating it (and any missing parent directories) if needed, "
            "overwriting if it exists. Path is relative to the working directory."
        ),
        parameters={
            "path": {"type": "string", "description": "Relative path to the file"},
            "content": {"type": "string", "description": "Text content to write"},
        },
        block=write_file,
    )

    def delete_file(*, path: str) -> str:
        target = resolve(path)
        if isinstance(target, str):
            return target
        if not target.is_file():
            return oops(f"'{path}' is not a file")
        try:
            target.unlink()
            return f"ok: deleted {path}"
        except OSError as e:
            return oops(str(e))

    registry.tool(
        "delete_file",
        description="Delete a file. Directories are not deleted. Path is relative to the working directory.",
        parameters={"path": {"type": "string", "description": "Relative path to the file to delete"}},
        block=delete_file,
    )

    def search_files(*, pattern: str, path: str = ".", glob: str = "*") -> str:
        target = resolve(path)
        if isinstance(target, str):
            return target

        files = [target] if target.is_file() else sorted(target.glob(f"**/{glob}"), key=str)

        try:
            regex = re.compile(pattern)
        except re.error as e:
            return oops(f"invalid pattern: {e}")

        matches: list[str] = []
        for file in files:
            if not file.is_file():
                continue
            rel = file.relative_to(root)
            try:
                for lineno, line in enumerate(file.read_text().splitlines(), start=1):
                    if regex.search(line):
                        matches.append(f"{rel}:{lineno}:{line}")
            except (OSError, UnicodeDecodeError) as e:
                matches.append(f"{rel}: error reading file: {e}")

        return "\n".join(matches) if matches else "no matches"

    registry.tool(
        "search_files",
        description=(
            "Search for a text pattern (literal string or regex) across all files in the working "
            "directory tree. Returns matching lines in 'path:line_number:content' format."
        ),
        parameters={
            "pattern": {"type": "string", "description": "The text or regex pattern to search for"},
            "path": {
                "type": "string",
                "description": "Subdirectory or file to search within (default '.' = entire working directory)",
            },
            "glob": {
                "type": "string",
                "description": "File glob to restrict which files are searched, e.g. '*.py' (default '*')",
            },
        },
        block=search_files,
    )
